internal/server/recon/recognize.go

```diff
@@ -73,6 +77,189 @@ func RecognizePlaces(
+		queryRespItems := getItemsForSpans(spans, query, span2item)
+		resp.QueryItems[query] = &pb.RecognizePlacesResponse_Items{Items: queryRespItems}
+	}
+	return resp, nil
+}
+
+// Takes a query and list of spans and gets the corresponding list of recognize
+// places response items. With the list of spans, we want the response items to
+// have a subset of non-overlapping spans. We do that with a greedy approach of
+// matching the longest span in query, getting the remaining query parts, and
+// recursively doing the match.
+func getItemsForSpans(spans []string, query string, span2item map[string]*pb.RecognizePlacesResponse_Item) []*pb.RecognizePlacesResponse_Item {
+	queryRespItems := []*pb.RecognizePlacesResponse_Item{}
+	// If empty query, return empty list
+	if len(query) == 0 {
+		return queryRespItems
+	}
+	// If empty list of spans, return the query as the only item
+	if len(spans) == 0 {
+		return append(queryRespItems, &pb.RecognizePlacesResponse_Item{Span: query})
+	}
+	span := spans[0]
+	queryParts := splitQueryBySpan(query, span)
+	for _, part := range queryParts {
+		if part == span {
+			queryRespItems = append(queryRespItems, span2item[span])
+		} else {
+			queryRespItems = append(queryRespItems, getItemsForSpans(spans[1:], part, span2item)...)
+		}
+	}
+	return queryRespItems
+}
+
+// Splits a query by a span into a list of parts like: non-span part, span part,
+// non-span part, span part, etc. If the query does not contain the span,
+// returns a list with the query as the only item.
+// Examples
+// query: "ab cd ef ab g", span: "ab", result: ["ab", "cd ef", "ab", "g"]
+// query: "ab cd ef g", span: "jk", result: ["ab cd ef g"]
+func splitQueryBySpan(query string, span string) []string {
+	splits := strings.Split(query, span)
+	// if span is not found in query, return just the query
+	if len(splits) < 2 {
+		return []string{query}
+	}
+	queryParts := []string{}
+	// prefix to add to the next split to be processed
+	nextPartPrefix := ""
+	// go through all the splits and add the span between each split if it was a
+	// valid split (we only want to split on complete words)
+	for i, split := range splits {
+		queryPart := nextPartPrefix + split
+		// if this is the last item in the list, add the query part and finish
+		if i == len(splits)-1 {
+			queryParts = append(queryParts, strings.TrimSpace(queryPart))
+			break
+		}
+		// Whether or not to skip adding anything in this round. We want to skip if
+		// this was not a valid split.
+		shouldSkipAdding := false
+		// if the current split is not empty and does not end in a space, assume
+		// span is part of the last word in current split which is not a valid split.
+		if len(split) > 0 && !strings.HasSuffix(split, " ") {
+			shouldSkipAdding = true
+		}
+		// if the next split is not empty and does not start with a space, assume
+		// span is part of the first word in next split which is not a valid split.
+		if len(splits[i+1]) > 0 && !strings.HasPrefix(splits[i+1], " ") {
```

Comment:
Can this happen for the first split?  Example: "foo ba" is the span to search and splitting "foo bar foo ba r", we'll get:

"r ", " r"

Reply:
updated the algorithm and this shouldn't happen